src/base/USong.pas

```diff
@@ -796,281 +799,328 @@ function TSong.ReadTXTHeader(SongFile: TTextFileStream; ReadCustomTags: Boolean)
-      begin
-        DecodeStringUTF8(Value, Genre, Encoding);
-        self.GenreASCII := LowerCase(TransliterateToASCII(Genre));
-      end
-
-      //Edition Sorting
-      else if (Identifier = 'EDITION') then
-      begin
-        DecodeStringUTF8(Value, Edition, Encoding);
-        self.EditionASCII := LowerCase(TransliterateToASCII(Edition));
-      end
-
-      //Creator Tag
-      else if (Identifier = 'CREATOR') then
-      begin
-        DecodeStringUTF8(Value, Creator, Encoding);
-        self.CreatorASCII := LowerCase(TransliterateToASCII(Creator));
-      end
-
-      //Language Sorting
-      else if (Identifier = 'LANGUAGE') then
-      begin
-        DecodeStringUTF8(Value, Language, Encoding);
-        self.LanguageASCII := LowerCase(TransliterateToASCII(Language));
-      end
-
-      //Year Sorting
-      else if (Identifier = 'YEAR') then
-      begin
-        TryStrtoInt(Value, self.Year)
-      end
-
-      // Song Start
-      else if (Identifier = 'START') then
-      begin
-        self.Start := StrToFloatI18n( Value )
-      end
-
-      // Song Ending
-      else if (Identifier = 'END') then
-      begin
-        TryStrtoInt(Value, self.Finish)
-      end
-
-      // Resolution
-      else if (Identifier = 'RESOLUTION') then
-      begin
-        TryStrtoInt(Value, self.Resolution);
-        if (self.Resolution < 1) then begin
-          Log.LogError('Ignoring invalid resolution in song: ' + FullFileName);
-          self.Resolution := DEFAULT_RESOLUTION;
-        end
-      end
-
-      // Notes Gap
-      else if (Identifier = 'NOTESGAP') then
-      begin
-        TryStrtoInt(Value, self.NotesGAP)
-      end
+        Continue;
+      end;
 
-      // Relative Notes
-      else if (Identifier = 'RELATIVE') then
-      begin
-        if (UpperCase(Value) = 'YES') then
-          self.Relative := true;
-      end
+      //Read Identifier and Value
+      Identifier  := UpperCase(Trim(Copy(Line, 2, SepPos - 2))); //Uppercase is for Case Insensitive Checks
+      Value       := Trim(Copy(Line, SepPos + 1, Length(Line) - SepPos));
 
-      // File encoding
-      else if (Identifier = 'ENCODING') then
+      //Check the Identifier (If Value is given)
+      if (Length(Value) = 0) then
       begin
-        self.Encoding := ParseEncoding(Value, Ini.DefaultEncoding);
+        Log.LogInfo('Empty field "'+Identifier+'" in file ' + FullFileName,
+                     'TSong.ReadTXTHeader');
+        AddCustomTag(Identifier, '');
       end
-
-      // PreviewStart
-      else if (Identifier = 'PREVIEWSTART') then
+      else
       begin
-        self.PreviewStart := StrToFloatI18n( Value );
-        if (self.PreviewStart>0) then
-        begin
-          MedleyFlags := MedleyFlags or 1;
-          HasPreview := true;
+        try
+          TagMap.Add(Identifier, Value);
+        except
+          on E: EListError do
+            Log.LogInfo('Duplicate Tag "'+Identifier+'" in file ' + FullFileName + ' will be ignored.',
```

Comment:
During testing I found that this line is never logged for me. To test I've just duplicated the `#TITLE` line so my txt looks like:
```
#TITLE: some title
#TITLE: some title THE DUPLICATE
```

old behaviour = it doesn't log anything and in-game it uses whatever came last in the file
new behaviour = the song isn't loaded at all and Error.log has:
```
ERROR:  Reading headers from file failed. File incomplete or not Ultrastar txt?: /path/to/my.txt
ERROR:  AnalyseFile failed for "/path/to/my.txt".
```

That first line is coming from `function TSong.Analyse` in USong.pas.

I'm pretty sure this isn't the intended new behaviour? Looking at the code I would have expected it to log the info, ignore the duplicate, but still load the song? If yes, know that I do agree with this new behaviour, I just don't know why it's not catching the exception locally (I'm on Linux).

Another scenario the old code can handle but the new one struggles with is custom tags. I'm not sure IF we should allow these to have duplicates, but this bit of txt header:
```
COMMENT:some comment
COMMENT:another comment
```
works fine on the old code, and when I edit and save the song through the Editor, _both_ comment lines are kept (and in order, too!). In the new code it refuses to load the song during scanning, same error as above.
(there isn't any spec on what to do with duplicate non-USDX tags, but we should probably not touch them)


I'm not worried that the exception is going to cause significant slowdowns for people with massive song libraries.

Reply:
The error should be fixed now. Another unit defined a different `EListError`, so that the code did not catch the actual Exception type from the `fgl` unit I wanted to catch.

Reply:
Next up I will change the duplicate handling, so that the duplicate custom tags will work again.
Probably I will allow duplicated tags in general to prepare for [multi-valued headers](https://github.com/UltraStar-Deluxe/format/blob/main/spec.md#31-single-valued-and-multi-valued-headers) introduced in format specification 1.1.0

Reply:
My newest commit now also supports duplicated custom tags again. (It actually also removes the original duplication handling, which I just fixed a commit prior.)
The handling of duplicated (non custom) tags is now again more similar to the old behavior, where the last value specified in the txt file is used.
A future implementation of multi-valued (non custom) headers should now also be a bit simpler to achieve.

Reply:
Thanks for updating! I just re-tested the PR and it works as expected now. I'll fix up the whitespace later tonight and then merge it.